subsys/bluetooth/mesh/scheduler_srv.c

```diff
@@ -0,0 +1,630 @@
+
+	return true;
+}
+
+static bool set_minute(struct tm *sched_time,
+		       struct tm *current_local,
+		       struct bt_mesh_schedule_register *item,
+		       struct bt_mesh_time_srv *srv)
+{
+	if (item->minute < current_local->tm_min) {
+		return false;
+	}
+
+	bool ovflw = false;
+
+	if (item->minute == EVERY_15_UNITS) {
+		sched_time->tm_min = current_local->tm_min < 15 ? 15 :
+				current_local->tm_min < 30 ? 30 :
+				current_local->tm_min < 45 ? 45 : 0;
+		ovflw = current_local->tm_min == 0 ? true : false;
+	} else if (item->minute == EVERY_20_UNITS) {
+		sched_time->tm_min = current_local->tm_min < 20 ? 20 :
+				current_local->tm_min < 40 ? 40 : 0;
+		ovflw = current_local->tm_min == 0 ? true : false;
+	} else if (item->minute == RANDOM_UNIT) {
+		sched_time->tm_min = sys_rand32_get() % 60;
+		ovflw = true;
+	} else {
+		sched_time->tm_min = item->minute == ANY_UNIT ?
+				current_local->tm_min : item->minute;
+	}
+
+	if (ovflw) {
+		return revise_hour(sched_time, current_local, item, srv, 1);
+	}
+
+	return true;
+}
+
+static bool set_second(struct tm *sched_time,
+		       struct tm *current_local,
+		       struct bt_mesh_schedule_register *item,
+		       struct bt_mesh_time_srv *srv)
+{
+	if (item->second < current_local->tm_sec) {
+		return false;
+	}
+
+	bool ovflw = false;
+
+	if (item->second == EVERY_15_UNITS) {
+		sched_time->tm_sec = current_local->tm_sec < 15 ? 15 :
+				current_local->tm_sec < 30 ? 30 :
+				current_local->tm_sec < 45 ? 45 : 0;
+		ovflw = current_local->tm_sec == 0 ? true : false;
+	} else if (item->second == EVERY_20_UNITS) {
+		sched_time->tm_sec = current_local->tm_sec < 20 ? 20 :
+				current_local->tm_sec < 40 ? 40 : 0;
+		ovflw = current_local->tm_sec == 0 ? true : false;
+	} else if (item->second == RANDOM_UNIT) {
+		sched_time->tm_sec = sys_rand32_get() % 60;
+		ovflw = true;
+	} else {
+		sched_time->tm_sec = item->second == ANY_UNIT ?
+				current_local->tm_sec : item->second;
+	}
+
+	if (ovflw) {
+		return revise_minute(sched_time, current_local, item, srv, 1);
+	}
+
+	return true;
+}
+
+static void schedule_action(struct bt_mesh_scheduler_srv *srv,
+			    uint8_t index)
+{
+	struct tm sched_time;
+	struct bt_mesh_schedule_register *item = &srv->data.sch_reg[index];
+	struct bt_mesh_scheduler_scheduled_meta *metadata =
+			&srv->data.metadata[index];
+
+	int64_t current_uptime = k_uptime_get();
+	struct tm *current_local = bt_mesh_time_srv_localtime(srv->time_srv,
+			current_uptime);
+
+	if (!set_year(&sched_time, current_local, item)) {
+		return;
+	}
+
+	if (!set_month(&sched_time, current_local, item)) {
+		return;
+	}
+
+	if (!set_day(&sched_time, current_local, item, srv->time_srv)) {
+		return;
+	}
+
+	if (!set_hour(&sched_time, current_local, item, srv->time_srv)) {
+		return;
+	}
+
+	if (!set_minute(&sched_time, current_local, item, srv->time_srv)) {
+		return;
+	}
+
+	if (!set_second(&sched_time, current_local, item, srv->time_srv)) {
+		return;
+	}
```

Comment:
I will have to trust you on this algorithm I think. It'll need some pretty good test coverage.

Reply:
I fully agree with you. This should be covered by unit tests as much as possible. There are mant corner cases in time conversion and rescheduling.